File: AlgorithmsPackCPP/1136. Parallel Courses.cpp
class Solution {
    typedef vector<vector<int>>GRAPH;


    void addEdge(GRAPH&graph, int from, int to){
        graph[from].push_back(to);
    }

public:
    int topSort(GRAPH&graph){
        int nodes = graph.size();
        vector<int>incomingEdges(nodes);
        for(int i = 0; i < nodes; i++){
            for(int j:graph[i]){
                incomingEdges[j]++;
            }
        }
        queue<int>q;
        for(int i = 0; i < nodes; i++){
            if(incomingEdges[i]==0){
                q.push(i);
            }
        }
        int nodesFound = 0;
        int result = 0;
        while(!q.empty()){
            int sz = q.size();
            while(sz--){
                int cur = q.front();
                q.pop();
                nodesFound++;
                for(int j:graph[cur]){
                    if(--incomingEdges[j]==0)
                        q.push(j);
                }
            }
            result++;
        }
        if(nodesFound!=nodes)
            return -1;
        return result;
    }

    int minimumSemesters(int n, vector<vector<int>>& relations) {
        GRAPH graph(n);
        for(auto&rel:relations)
            addEdge(graph,rel[0]-1,rel[1]-1);

        return topSort(graph);
    }
};

File: AlgorithmsPackJavaSolutions/1136. Parallel Courses.java
class Solution {
    public int minimumSemesters(int n, int[][] relations) {
        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            graph.add(new ArrayList<>());
        }

        for (int[] rel : relations) {
            addEdge(graph, rel[0] - 1, rel[1] - 1);
        }

        return topSort(graph);
    }

    private void addEdge(List<List<Integer>> graph, int from, int to) {
        graph.get(from).add(to);
    }

    private int topSort(List<List<Integer>> graph) {
        int nodes = graph.size();
        int[] incomingEdges = new int[nodes];
        for (int i = 0; i < nodes; i++) {
            for (int j : graph.get(i)) {
                incomingEdges[j]++;
            }
        }

        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < nodes; i++) {
            if (incomingEdges[i] == 0) {
                q.offer(i);
            }
        }

        int nodesFound = 0;
        int result = 0;
        while (!q.isEmpty()) {
            int sz = q.size();
            while (sz > 0) {
                int cur = q.poll();
                nodesFound++;
                for (int j : graph.get(cur)) {
                    incomingEdges[j]--;
                    if (incomingEdges[j] == 0) {
                        q.offer(j);
                    }
                }
                sz--;
            }
            result++;
        }

        if (nodesFound != nodes) {
            return -1;
        }
        return result;
    }
}
